recipes/full_finetune_single_device.py

```diff
@@ -221,18 +229,47 @@ def _setup_model(
         return model
 
     def _setup_optimizer(
-        self, cfg_optimizer: DictConfig, opt_state_dict: Optional[Dict[str, Any]] = None
-    ) -> Optimizer:
+        self,
+        cfg_optimizer: DictConfig,
+        optimizer_in_bwd: bool = False,
+        opt_state_dict: Optional[Dict[str, Any]] = None,
+    ) -> Union[Optimizer, None]:
         """
         Set up the optimizer. This method also handles loading the optimizer state_dict, if specified.
         """
-        optimizer = config.instantiate(cfg_optimizer, self._model.parameters())
-
-        if opt_state_dict:
-            optimizer.load_state_dict(opt_state_dict)
-
-        log.info("Optimizer is initialized.")
-        return optimizer
+        if optimizer_in_bwd:
+            # Maintain a dict of optims for every parameter.
+            # TODO (rohan-varma): check foreach arg
```

Comment:
I don't understand what this means in this context. Isn't foreach kind of a no-op in this case?

Reply:
You're right! Foreach should not really matter here - might just run a quick test to confirm.